core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/chunkedencoding/ChunkEncodedInputStream.java

```diff
@@ -0,0 +1,162 @@
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.aws.chunkedencoding;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.http.auth.aws.internal.io.SdkInputStream;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Pair;
+
+
+/**
+ * An implementation of chunked transfer encoding, which wrap an @{link InputStream}.
+ * This implementation supports chunk-headers, chunk-extensions, and trailers.
+ */
+@SdkProtectedApi
+public final class ChunkEncodedInputStream extends SdkInputStream {
+    private static final Logger LOG = Logger.loggerFor(ChunkEncodedInputStream.class);
+    private static final byte[] CRLF = {'\r', '\n'};
+    private static final byte[] END = {};
+
+    private final InputStream inputStream;
+    private final int chunkSize;
+
+    private final ChunkHeader header;
+    private final List<ChunkExtension> extensions = new ArrayList<>();
+    private final List<Trailer> trailers = new ArrayList<>();
+
+    private Chunk currentChunk;
+    private boolean isFinished = false;
+
+    public ChunkEncodedInputStream(
+        InputStream inputStream,
+        int chunkSize,
+        ChunkHeader header,
+        List<ChunkExtension> extensions,
+        List<Trailer> trailers
+    ) {
+        this.inputStream = inputStream;
+        this.chunkSize = chunkSize;
+        this.header = header;
+        this.extensions.addAll(extensions);
+        this.trailers.addAll(trailers);
+    }
+
+    @Override
+    protected InputStream getWrappedInputStream() {
+        return inputStream;
+    }
+
+    public int getChunkSize() {
+        return chunkSize;
+    }
+
+    @Override
+    public int read() throws IOException {
+        if (currentChunk == null || currentChunk.ended() && !isFinished) {
+            readChunk(inputStream);
+        }
+
+        return currentChunk.stream().read();
+    }
+
+    private void readChunk(InputStream stream) throws IOException {
+        LOG.debug(() -> "Reading next chunk.");
+        if (currentChunk != null) {
+            currentChunk.close();
+        }
+        // we *have* to read from the backing stream in order to figure out if it's the end or not
+        byte[] chunkData = new byte[chunkSize];
+        int read = stream.read(chunkData, 0, chunkSize);
+
+        if (read > 0) {
+            // form the new chunk, which includes the header, the extensions, and the chunk data
+            byte[] chunk = Arrays.copyOf(chunkData, read);
```

Comment:
I was a bit confused reading this code, because we create a chunk, send it to the output stream and then assign the result back to the chunk. From a readability POV, can we inline the array copy to the writeChunk call?